src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs

```diff
@@ -1,38 +1,212 @@
-using System;
+﻿using System;
+using System.Collections.Generic;
 using System.Linq;
 using System.Linq.Expressions;
+using System.Reflection;
+using System.Runtime.CompilerServices;
 using NHibernate.Collection;
 using NHibernate.Util;
-using Remotion.Linq.Clauses.Expressions;
 using Remotion.Linq.Parsing;
-using Remotion.Linq.Parsing.ExpressionVisitors;
 using Remotion.Linq.Parsing.ExpressionVisitors.TreeEvaluation;
 
 namespace NHibernate.Linq.Visitors
 {
-	internal class NhPartialEvaluatingExpressionVisitor : RelinqExpressionVisitor, IPartialEvaluationExceptionExpressionVisitor
+	// Copied from Relinq and added logic for detecting and linking variables with evaluated constant expressions
+	/// <summary>
+	/// Takes an expression tree and first analyzes it for evaluatable subtrees (using <see cref="EvaluatableTreeFindingExpressionVisitor"/>), i.e.
+	/// subtrees that can be pre-evaluated before actually generating the query. Examples for evaluatable subtrees are operations on constant
+	/// values (constant folding), access to closure variables (variables used by the LINQ query that are defined in an outer scope), or method
+	/// calls on known objects or their members. In a second step, it replaces all of the evaluatable subtrees (top-down and non-recursive) by 
+	/// their evaluated counterparts.
+	/// </summary>
+	/// <remarks>
+	/// This visitor visits each tree node at most twice: once via the <see cref="EvaluatableTreeFindingExpressionVisitor"/> for analysis and once
+	/// again to replace nodes if possible (unless the parent node has already been replaced).
+	/// </remarks>
+	internal sealed class NhPartialEvaluatingExpressionVisitor : RelinqExpressionVisitor
```

Comment:
Unfortunatelly, Relinq `PartialEvaluatingExpressionVisitor` class is sealed and the only way to link the evaluated expression with the variable was to copy the class and add the required logic.

Reply:
Added the license and copyright from Relinq.